---
title: "G2G_TB_Report"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library("ggVennDiagram")
library("ggplot2")
library(EnvStats)
library(ggpubr)
library(networkD3)
library(RColorBrewer)

```

## Sample Status - Raw

```{r stats}
Genotyped_Samples <- data.table::fread('../../data/Genotyping/TB_DAR_Genotyping/Filt_Merged/TBDAR.Genotyped.Merged.Filt.fam')$V2
Genotyped_Samples <- sapply(Genotyped_Samples,function(x) gsub(x=x,pattern = 'Batch1_',replacement = ''))
Genotyped_Samples <- sapply(Genotyped_Samples,function(x) gsub(x=x,pattern = 'Batch2_',replacement = ''))

WGS_Samples <- data.table::fread('../../data/Genotyping_WGS/TBDAR.WGS.Imputed.GWASReady.fam')$V2
WGS_Samples <- WGS_Samples[sapply(WGS_Samples,function(x) grepl(pattern = 'WGS',x=x))]
WGS_Samples <- sapply(WGS_Samples,function(x) gsub(x=x,pattern = 'WGS.Fellay.',replacement = ''))

Mtb_Samples <- unique(sapply(dir('../../data/Mtb/Mtb_VCF_files/full_allpos/'),function(x) strsplit(x=x,split = '\\.')[[1]][1]))
Metadata <- data.table::fread('../../data/pheno/metadata_Sinergia_final_dataset_human_bac_genome_available_corrected.txt') %>% dplyr::select(PATIENT_ID,G_NUMBER)
Mtb_Only <- setdiff(Mtb_Samples,Metadata$G_NUMBER)

Mtb_Patient_IDs <- c(Mtb_Only,Metadata$PATIENT_ID)

ggVennDiagram(list(Genotyped_Samples,WGS_Samples,Mtb_Patient_IDs),category.names = c("Host\nGenotyped","Host\nWGS",'Mtb\nWGS'))


```
## QC - Regenotyped
```{r regenotyped}
TB_DAR_Samples <- system('~/Software/bcftools query -l ../../data/Genotyping_WGS/TBDAR.WGS.Imputed.AnalysisReady.vcf.gz',intern = T)
Patient_IDs <- sapply(TB_DAR_Samples,function(x) strsplit(x,split = '_')[[1]][2])
Patient_IDs <- sapply(Patient_IDs,function(x) gsub(x=x,pattern = 'Fellay.',replacement = ''))

Dup_IDs <- TB_DAR_Samples[which(Patient_IDs %in% Patient_IDs[duplicated(Patient_IDs)])]
Dup_IDs_df <- data.frame(FID = Dup_IDs,IID = Dup_IDs)
data.table::fwrite(Dup_IDs_df,'../../data/Genotyping_WGS/regenotyped.ids',sep = ' ',row.names = F,quote = F)
system('~/Software/plink2 --bfile ../../data/Genotyping_WGS/TBDAR.WGS.Imputed.AnalysisReady --chr 1-22 --keep ../../data/Genotyping_WGS/regenotyped.ids --make-bed --out ../../data/Genotyping_WGS/TBDAR.Regenotyped')

regenotyped_plink <- snpStats::read.plink('../../data/Genotyping_WGS/TBDAR.Regenotyped')

WGS_Genotypes <- regenotyped_plink$genotypes[sapply(regenotyped_plink$fam$member,function(x) grepl(x=x,pattern = 'WGS')),]
rownames(WGS_Genotypes) <- sapply(rownames(WGS_Genotypes) ,function(x) strsplit(x=x,split = '\\.')[[1]][2])

H3Africa_Genotypes <- regenotyped_plink$genotypes[!sapply(regenotyped_plink$fam$member,function(x) grepl(x=x,pattern = 'WGS')),]
H3Africa_Genotypes <- H3Africa_Genotypes[!duplicated(sapply(rownames(H3Africa_Genotypes),function(x) strsplit(x=x,split = '_')[[1]][2])),]
rownames(H3Africa_Genotypes) <- sapply(rownames(H3Africa_Genotypes),function(x) strsplit(x=x,split = '_')[[1]][2])
H3Africa_Genotypes <- H3Africa_Genotypes[rownames(WGS_Genotypes),]

snp_comp <- snpStats::sm.compare(H3Africa_Genotypes,WGS_Genotypes,col.wise = T)

sample_diff <- (snp_comp$row.wise[,'Disagree'] - snp_comp$row.wise[,'NA.disagree']) / (snp_comp$row.wise[,'Agree'] + snp_comp$row.wise[,'Disagree'] - snp_comp$row.wise[,'NA.disagree'])
snp_diff <- (snp_comp$col.wise[,'Disagree'] - snp_comp$col.wise[,'NA.disagree']) / (snp_comp$col.wise[,'Agree'] + snp_comp$col.wise[,'Disagree'] - snp_comp$col.wise[,'NA.disagree'])
hist(sample_diff,main = 'Discordant SNPs per Sample',xlab = 'Fraction')
#Exclude duplicate between batch 1 and batch2. Exclude WGS Samples
WGS_Dup_IDs <- dplyr::filter(Dup_IDs_df,grepl(x = IID,pattern = 'WGS'))
H3Africa_Dup_IDs <- dplyr::filter(Dup_IDs_df,!grepl(x = IID,pattern = 'WGS'))
H3Africa_Dup_IDs$Simple_ID <- sapply(H3Africa_Dup_IDs$IID,function(x) strsplit(x=x,split = '_')[[1]][2])

#Find mismatch Duplicates
mismatching_Dup_IDs <- regenotyped_plink$fam$member[sapply(regenotyped_plink$fam$member,function(x) any(sapply(names(sample_diff[sample_diff > 0.1]),function(y) grepl(x=x,pattern = y))))]

```

## QC - Sex Mismatch
```{r check_sex}
tb_dar_pheno <- data.table::fread('../../data/pheno/metadata_Sinergia_final_all_pats.txt') %>% dplyr::mutate(PATIENT_ID = as.character(PATIENT_ID))

WGS_Sex_Check <- data.table::fread('../../data/WGS/WGS_Host_Data/plink.sexcheck') %>% dplyr::select(PLINK_ID=IID,SNPSEX) %>% dplyr::mutate(Source = 'WGS')
WGS_Sex_Check$SNPSEX <- ifelse(WGS_Sex_Check$SNPSEX == 1, 'MALE','FEMALE')
WGS_Sex_Check$PATIENT_ID <- sapply(WGS_Sex_Check$PLINK_ID,function(x) gsub(x=x,pattern = 'WGS_Fellay.',replacement = ''))

Batch1_Sex_Check <- data.table::fread('../../data/Genotyping/TB_DAR_Genotyping/Batch1/Sex_Est_Table.txt') %>% dplyr::select(PATIENT_ID=`Sample ID`,SNPSEX = Gender) %>% dplyr::mutate(SNPSEX = toupper(SNPSEX),Source = 'Batch1')
Batch1_Sex_Check$PATIENT_ID <- sapply(Batch1_Sex_Check$PATIENT_ID,function(x) strsplit(x=x,split = '-')[[1]][1])
Batch1_Sex_Check$PATIENT_ID <- sapply(Batch1_Sex_Check$PATIENT_ID,function(x) gsub(x=x,pattern = ' V01 B1',replacement = ''))
Batch1_Sex_Check$PLINK_ID <- paste0('Batch1_',Batch1_Sex_Check$PATIENT_ID)

Batch2_Sex_Check <- data.table::fread('../../data/Genotyping/TB_DAR_Genotyping/Batch2/Sex_Est_Table.txt') %>% dplyr::select(PATIENT_ID=`Sample ID`,SNPSEX = Gender) %>% dplyr::mutate(SNPSEX = toupper(SNPSEX),Source = 'Batch2')
Batch2_Sex_Check$PATIENT_ID <- sapply(Batch2_Sex_Check$PATIENT_ID,function(x) strsplit(x=x,split = '-')[[1]][1])
Batch2_Sex_Check$PATIENT_ID <- sapply(Batch2_Sex_Check$PATIENT_ID,function(x) gsub(x=x,pattern = 'TZ ',replacement = ''))
Batch2_Sex_Check$PLINK_ID <- paste0('Batch2_',Batch2_Sex_Check$PATIENT_ID)

sex_comparison <- dplyr::left_join(rbind(WGS_Sex_Check,Batch1_Sex_Check,Batch2_Sex_Check),tb_dar_pheno %>% dplyr::select(patient_sex,PATIENT_ID) %>% dplyr::mutate(patient_sex = toupper(patient_sex)),by = c('PATIENT_ID'='PATIENT_ID'))

sex_mismatch <- dplyr::filter(sex_comparison,SNPSEX != patient_sex & SNPSEX != 'UNKNOWN')

```
## QC - Genetic Duplicates
```{r genetic_dup}
king_dup_full <- data.table::fread('../../data/Genotyping_WGS/TBDAR.WGS.Imputed.AnalysisReady.kin0') %>% 
  dplyr::filter(InfType == 'Dup/MZ') %>% dplyr::select(ID1,ID2)
king_dup_full$ID1_Simple <- sapply(king_dup_full$ID1,function(x) gsub(x=x,pattern = 'Batch1_',replacement = ''))
king_dup_full$ID1_Simple <- sapply(king_dup_full$ID1_Simple,function(x) gsub(x=x,pattern = 'Batch2_',replacement = ''))
king_dup_full$ID1_Simple <- sapply(king_dup_full$ID1_Simple,function(x) gsub(x=x,pattern = 'WGS_Fellay.',replacement = ''))

king_dup_full$ID2_Simple <- sapply(king_dup_full$ID2,function(x) gsub(x=x,pattern = 'Batch1_',replacement = ''))
king_dup_full$ID2_Simple <- sapply(king_dup_full$ID2_Simple,function(x) gsub(x=x,pattern = 'Batch2_',replacement = ''))
king_dup_full$ID2_Simple <- sapply(king_dup_full$ID2_Simple,function(x) gsub(x=x,pattern = 'WGS_Fellay.',replacement = ''))

genetic_duplicates <- dplyr::filter(king_dup_full,ID1_Simple != ID2_Simple)


```
## QC - Relatives

```{r relatives}
relatives <- data.table::fread('../../data/Genotyping_WGS/TBDAR.WGS.Imputed.AnalysisReady.kin0') %>% 
  dplyr::filter(InfType != 'Dup/MZ') %>% dplyr::select(ID1,ID2,InfType)
relatives$ID1_Simple <- sapply(relatives$ID1,function(x) gsub(x=x,pattern = 'Batch1_',replacement = ''))
relatives$ID1_Simple <- sapply(relatives$ID1_Simple,function(x) gsub(x=x,pattern = 'Batch2_',replacement = ''))
relatives$ID1_Simple <- sapply(relatives$ID1_Simple,function(x) gsub(x=x,pattern = 'WGS_Fellay.',replacement = ''))

relatives$ID2_Simple <- sapply(relatives$ID2,function(x) gsub(x=x,pattern = 'Batch1_',replacement = ''))
relatives$ID2_Simple <- sapply(relatives$ID2_Simple,function(x) gsub(x=x,pattern = 'Batch2_',replacement = ''))
relatives$ID2_Simple <- sapply(relatives$ID2_Simple,function(x) gsub(x=x,pattern = 'WGS_Fellay.',replacement = ''))

relatives_pairs <- lapply(1:nrow(relatives),function(i) sort(c(relatives$ID1_Simple[i],relatives$ID2_Simple[i])))
relatives_pairs <- relatives_pairs[!duplicated(relatives_pairs)]
relatives_pairs_list <- data.frame(from = sapply(relatives_pairs,function(x) x[1]),to = sapply(relatives_pairs,function(x) x[2]))
#Manually check to see if there are trios 
p <- simpleNetwork(relatives_pairs_list, height="50px", width="50px")

```
## QC - PCA
```{PCA}
GetPCAOutliers <- function(PCA_results,target_POP){
  pop_tbl <- data.table::fread('~/1KG_PCA/1KG_data/20130606_g1k.ped') %>% dplyr::select(ID=`Individual ID`,Population)
  pop_info <- data.table::fread('~/1KG_PCA/1KG_data/20131219.populations.tsv') %>% dplyr::select(Population = `Population Code`,SuperPopulation = `Super Population`)
  pop_tbl <- dplyr::left_join(pop_tbl,pop_info)
  
  genotyping_pca <- data.table::fread(PCA_results) %>% dplyr::select(ID=V2,PC1=V3,PC2=V4)
  genotyping_pca <- genotyping_pca %>% dplyr::left_join(pop_tbl %>% dplyr::select(ID,SuperPopulation))
  
  genotyping_pca$SuperPopulation[is.na(genotyping_pca$SuperPopulation)] <- 'TARGET'
  
  library(class)
  non_target_set <- genotyping_pca %>% dplyr::filter(SuperPopulation != 'TARGET' & SuperPopulation != 'AMR') %>% dplyr::select(-SuperPopulation,-ID) #Train KNN based on 1KG 
  non_target_set_class <- genotyping_pca %>% dplyr::filter(SuperPopulation != 'TARGET' & SuperPopulation != 'AMR') %>% dplyr::select(SuperPopulation) #Labels for the Train set
  non_target_set_class <- factor(non_target_set_class$SuperPopulation)
  
  train_sample <- sample(1:nrow(non_target_set),size = 0.7*nrow(non_target_set),replace = F)
  train_set <- non_target_set[train_sample,]
  train_set_class <- non_target_set_class[train_sample]
  test_set <- non_target_set[-train_sample,]
  
  target_set <- genotyping_pca %>% dplyr::filter(SuperPopulation == 'TARGET') 
  target_set_pred <- target_set  %>% dplyr::select(-ID,-SuperPopulation)
  knn_test <- knn(train = train_set,test = test_set,cl=train_set_class,k=4)
  print(glue::glue('Testing Accuracy: {sum(knn_test == non_target_set_class[-train_sample]) / length(knn_test)}'))
  
  knn_target <- knn(train = non_target_set,test = target_set_pred,cl=non_target_set_class,k=4) 
  outliers <- target_set$ID[knn_target != target_POP]
  p1 <- ggplot2::ggplot(genotyping_pca %>% dplyr::filter(SuperPopulation != 'AMR')) + aes(x=PC1,y=PC2,color = SuperPopulation) + geom_point() + ggrepel::geom_text_repel(data = genotyping_pca %>% dplyr::filter(ID %in%outliers),aes(x=PC1,y=PC2,label = ID))
  
  return(list(outliers=outliers,p1=p1))
  
}
Genotyping_WGS_outliers <- GetPCAOutliers('../../data/Genotyping_WGS/PCA/TBDAR.WGS.Imputed.eigenvec','AFR')
```

## Sample Status - After QC

```{r stats_after_QC}
Host_Samples <- data.table::fread('../../data/Genotyping_WGS/TBDAR.WGS.Imputed.GWASReady.fam')$V2

WGS_Samples <- Host_Samples[sapply(Host_Samples,function(x) grepl(pattern = 'WGS',x=x))]
WGS_Samples <- sapply(WGS_Samples,function(x) gsub(x=x,pattern = 'WGS.Fellay.',replacement = ''))

Genotyped_Samples <- Host_Samples[sapply(Host_Samples,function(x) grepl(pattern = 'Batch',x=x))]
Genotyped_Samples <- sapply(Genotyped_Samples,function(x) gsub(x=x,pattern = 'Batch1_',replacement = ''))
Genotyped_Samples <- sapply(Genotyped_Samples,function(x) gsub(x=x,pattern = 'Batch2_',replacement = ''))

Mtb_Samples <- unique(sapply(dir('../../data/Mtb/Mtb_VCF_files/full_allpos/'),function(x) strsplit(x=x,split = '\\.')[[1]][1]))
Metadata <- data.table::fread('../../data/pheno/metadata_Sinergia_final_dataset_human_bac_genome_available.txt') %>% dplyr::select(PATIENT_ID,G_NUMBER) %>% dplyr::filter(PATIENT_ID %in% c(Genotyped_Samples,WGS_Samples))
Mtb_Only <- setdiff(Mtb_Samples,Metadata$G_NUMBER)

Mtb_Patient_IDs <- c(Mtb_Only,Metadata$PATIENT_ID)

ggVennDiagram(list(Genotyped_Samples,WGS_Samples,Mtb_Patient_IDs),category.names = c("Host\nGenotyped","Host\nWGS",'Mtb\nWGS'))


```
## GWAS
```{r GWAS}
TB_Score_GWAS <- data.table::fread('~/TB_GWAS/GWAS/TB_score.TB_score.glm.linear')
p1 <- ggman::ggman(TB_Score_GWAS %>% dplyr::filter(P<0.1),snp = "ID", bp = "POS", chrom = "#CHROM", pvalue = "P") + ggtitle('TB Score') + ylim(1,10) 

L1_n3_GWAS <- data.table::fread('~/TB_GWAS/GWAS/PC_3/introduction.LineageL1_n3.glm.logistic.hybrid')
L2_n1568_GWAS <- data.table::fread('~/TB_GWAS/GWAS/PC_3/introduction.LineageL2_n1568.glm.logistic.hybrid')
L3_n1111_GWAS <- data.table::fread('~/TB_GWAS/GWAS/PC_3/introduction.LineageL3_n1111.glm.logistic.hybrid')
L4_n296_GWAS <- data.table::fread('~/TB_GWAS/GWAS/PC_3/introduction.LineageL4_n296.glm.logistic.hybrid')
p6 <- ggman::ggman(L1_n3_GWAS %>% dplyr::filter(P<0.1),snp = "ID", bp = "POS", chrom = "#CHROM", pvalue = "P") + ggtitle('L1_n3') + ylim(1,10) 
p7 <- ggman::ggman(L2_n1568_GWAS %>% dplyr::filter(P<0.1),snp = "ID", bp = "POS", chrom = "#CHROM", pvalue = "P") + ggtitle('L2_n1568') + ylim(1,10) 
p8 <- ggman::ggman(L3_n1111_GWAS %>% dplyr::filter(P<0.1),snp = "ID", bp = "POS", chrom = "#CHROM", pvalue = "P") + ggtitle('L3_n1111') + ylim(1,10) 
p9 <- ggman::ggman(L4_n296_GWAS %>% dplyr::filter(P<0.1),snp = "ID", bp = "POS", chrom = "#CHROM", pvalue = "P") + ggtitle('L4_n296') + ylim(1,10) 
ggpubr::ggarrange(p6,p7,p8,p9,nrow=4)


```


## G2G Results - Manhattan
```{r manhattan}
fixA_res <- data.table::fread(cmd = 'zcat ../../results/Burden_False_SIFT_False_Del_False_HomoOnly_True_HetThresh_10/PLINK/PC_3_pPC_0/Stratified_False/LINEAGE_ALL/fixA_Rv3029c:3388671:p.Thr67Met.fixA_Rv3029c:3388671:p.Thr67Met.glm.logistic.hybrid.gz')
p1 <- ggman::ggman(fixA_res %>% dplyr::filter(P<0.1),snp = "ID", bp = "POS", chrom = "#CHROM", pvalue = "P") + ggtitle('fixA_Rv3029c:3388671:p.Thr67Met') + geom_hline(yintercept = -log10(5e-8/420),colour = 'red',linetype  = 'dashed') + ylim(1,12) 
fixA_sig <- dplyr::filter(fixA_res,P < 5e-8/519)
ggman::ggmanLabel(p1, labelDfm = fixA_sig, snp = "ID", label = "ID")
GWAS.utils::QQplot(fixA_res$P,main = "fixA_Rv3029c:3388671:p.Thr67Met")
text(paste0('Lambda=',round(GWAS.utils::genomic_inflation(P=fixA_res$P),2)),x=1,y = 8)

rv2348_res <- data.table::fread(cmd = 'zcat ../../results/Burden_False_SIFT_False_Del_False_HomoOnly_True_HetThresh_10/PLINK/PC_3_pPC_0/Stratified_False/LINEAGE_ALL/Rv2348c_Rv2348c:2626678:p.Ile101Met.Rv2348c_Rv2348c:2626678:p.Ile101Met.glm.logistic.hybrid.gz')
p2 <- ggman::ggman(rv2348_res %>% dplyr::filter(P<0.1),snp = "ID", bp = "POS", chrom = "#CHROM", pvalue = "P") + ggtitle('Rv2348c_Rv2348c:2626678:p.Ile101Met') + geom_hline(yintercept = -log10(5e-8/420),colour = 'red',linetype  = 'dashed') + ylim(1,12) 
rv2348_sig <- dplyr::filter(rv2348_res,P < 5e-8/519)
ggman::ggmanLabel(p2, labelDfm = rv2348_sig, snp = "ID", label = "ID")
GWAS.utils::QQplot(rv2348_res$P,main = "Rv2348c_Rv2348c:2626678:p.Ile101Met")
text(paste0('Lambda=',round(GWAS.utils::genomic_inflation(P=rv2348_res$P),2)),x=1,y = 8)


# rv1149_res <- data.table::fread(cmd = 'zcat ../../results/Burden_True_SIFT_True_Del_True_HomoOnly_True_HetThresh_10/PLINK/PC_3_pPC_0/Stratified_False/LINEAGE_ALL/Rv1149_Rv1149.Rv1149_Rv1149.glm.logistic.hybrid.gz')
# p2 <- ggman::ggman(rv1149_res %>% dplyr::filter(P<0.1),snp = "ID", bp = "POS", chrom = "#CHROM", pvalue = "P") + ggtitle('Rv1149_Rv1149') + geom_hline(yintercept = -log10(5e-8/420),colour = 'red',linetype  = 'dashed') + ylim(1,14) 
# rv1149_sig <- dplyr::filter(rv1149_res,P < 5e-8/420)
# ggman::ggmanLabel(p2, labelDfm = rv1149_sig, snp = "ID", label = "ID")
# GWAS.utils::QQplot(rv1149_res$P,main = "Rv1149_Rv1149")
# text(paste0('Lambda=',round(GWAS.utils::genomic_inflation(P=rv1149_res$P),2)),x=1,y = 8)

```

## G2G Plot
```{r G2G_Plot}
library(dplyr)
library(GenomicFeatures)
library(ggplot2)
library(latex2exp)
library(ggrepel)
library(qqman)

#Pairwise Plot
G2G_Res <- readRDS('../../results/Burden_False_SIFT_False_Del_False_HomoOnly_True_HetThresh_10/PLINK/PC_3_pPC_0/Stratified_False/G2G_results.rds')
#Parse G2G Results into dataframe
G2G_Res <- G2G_Res$ALL
G2G_Res <- G2G_Res[sapply(G2G_Res,function(x) nrow(x) > 0)]

G2G_Res_Expand <- data.frame(Host_Chr = character(),Host_Pos = numeric(),Mtb_Pos = numeric(),Mtb_Gene=character(),P = numeric())
for(i in 1:length(G2G_Res)){
  cur_mtb_variant <- names(G2G_Res)[i]
  Host_Pos <- as.numeric(G2G_Res[[i]]$POS)
  Host_Chr  <- G2G_Res[[i]]$`#CHROM`
  Mtb_Pos <- as.numeric(strsplit(names(G2G_Res)[i],split = ':')[[1]][2])
  Mtb_Gene <- strsplit(names(G2G_Res)[i],split = ':')[[1]][1]
  P <- G2G_Res[[i]]$P

  G2G_Res_Expand <- rbind(G2G_Res_Expand,data.frame(Host_Chr=Host_Chr,Host_Pos=Host_Pos,Mtb_Pos=Mtb_Pos,Mtb_Gene=Mtb_Gene,P = P))
}

#Host Chr Length
host_chr_length <- getChromInfoFromUCSC("hg19")
host_chr_length <- dplyr::filter(host_chr_length,chrom %in% paste0('chr',c(1:22,'X'))) %>%
  dplyr::select(chr=chrom,size) %>%
  dplyr::mutate(chr = gsub(chr,pattern = 'chr',replacement = ''))
#Target Length
target_chr_length <- data.frame(chr='H37Rv',size = 4411532)

scale_factor <- sum(host_chr_length$size)/target_chr_length$size
target_chr_length <- data.frame(chr='H37Rv',size = 4411532*scale_factor)

#Change to Block Sturcture (For draw pairwise)
G2G_Block <- G2G_Res_Expand %>% dplyr::select(chr = Host_Chr,start = Host_Pos,tarSt=Mtb_Pos,P) %>%
  dplyr::mutate(orient=1,tarChr='H37Rv',tarSpecies = 'Mtb',tarSt=tarSt*scale_factor)


draw.pairwise <- function(dataTMP,ref_sizes,tar_sizes,refName,tarName,p_thresh) {
  xstart<-refchr<-tarchr<-x<-y<-group<-fill<-NULL
  data <- dplyr::select(dataTMP,tarchr=tarChr,tarstart=tarSt,refchr=chr,refstart=start,P)
  colnames(ref_sizes) = c("refchr", "size")
  colnames(tar_sizes) = c("tarchr", "size")

  #This adds gap in between reference chromosomes and convert to "linear" genome
  for (i in c(1:nrow(ref_sizes))){
    #print(i)
    if (i == 1){
      total_start = 1
      total_end = ref_sizes[i, "size"]
    } else {
      total_start = total_end + 6000000
      total_end = total_start + ref_sizes[i, "size"]
    }
    ref_sizes[i,"xstart"] = total_start
    ref_sizes[i, "xend"] = total_end
  }

  #This adds gap in between target chromosomes
  for (i in c(1:nrow(tar_sizes))){
    #print(i)
    if (i == 1){
      total_start = 1
      total_end = tar_sizes[i, "size"]
    } else {
      total_start = total_end + 6000000
      total_end = total_start + tar_sizes[i, "size"]
    }
    tar_sizes[i,"xstart"] = total_start
    tar_sizes[i, "xend"] = total_end
  }

  #This converts coordinates to linear genome and creates synteny polygon coordinates
  synteny = data.frame()
  for (i in c(1:nrow(data))){
    tar_chr = data[i,"tarchr"]
    ref_chr = data[i,"refchr"]
    dir = data[i, "dir"]

    tar_add = tar_sizes[as.character(tar_sizes$tarchr)==as.character(tar_chr),]$xstart
    ref_add = ref_sizes[as.character(ref_sizes$refchr)==as.character(ref_chr),]$xstart
    tar_y = 0.1
    ref_y = 2
    tar_xstart = data[i,"tarstart"] + tar_add
    ref_xstart = data[i,"refstart"] + ref_add

    df = data.frame(x = ref_xstart,y=ref_y,xend=tar_xstart,yend=tar_y,Sig = as.factor(ifelse(data$P[i] < p_thresh,'Sig','NonSig')))
    synteny = rbind(synteny,df)
  }

  #making sure chr columns are factors
  tar_sizes$tarchr<-as.factor(tar_sizes$tarchr)
  ref_sizes$refchr<-as.factor(ref_sizes$refchr)


  #This prints plot
  print(ggplot2::ggplot(size = 0.5, font = 10, data = data) +
          ggplot2::geom_rect(data=ref_sizes, mapping=ggplot2::aes(xmin=xstart, xmax=xend, ymin=2, ymax=2.10, fill=refchr),
                             color="black", alpha = 0.85, size = 0.2 ) +
          ggplot2::geom_text(data=ref_sizes,ggplot2::aes(x=(xstart+xend)/2,y=2.25,label=refchr),size=4,angle = 45) +
          ggplot2::geom_text(mapping=ggplot2::aes(x=10,y=2.4, label=refName),size=5) +
          ggplot2::geom_rect(data=tar_sizes, mapping=ggplot2::aes(xmin=xstart, xmax=xend, ymin=0, ymax=0.10),fill="purple",
                             color="black", alpha = 0.85, size = 0.2 ) +
          ggplot2::geom_text(data=tar_sizes,ggplot2::aes(x=(xstart+xend)/2,y=-0.2,label=tarchr),size=4) +
          ggplot2::geom_text(mapping=ggplot2::aes(x=10,y=-0.05, label=tarName),size=5) +
          ggplot2::geom_segment(data = synteny, ggplot2::aes(x = x, y = y,xend=xend,yend=yend,colour=Sig,alpha=Sig)) +
          ggplot2::scale_fill_manual(values = c("1" = "#BFD73B", "2" = "#39ACE2", "3" = "#F16E8A",
                                                "4" = "#2DB995", "5" = "#855823", "6" = "#A085BD",
                                                "7" = "#2EB560", "8" = "#D79128", "9" = "#FDBB63",
                                                "10" = "#AFDFE5", "11" = "#BF1E2D", "12" = "purple4",
                                                "13"= "#B59F31", "14" = "#F68B1F", "15" = "#EF374B",
                                                "16" = "#D376FF", "17" = "#009445", "18" = "#CE4699",
                                                "19" = "#7C9ACD", "20" = "#84C441", "21" = "#404F23",
                                                "22" = "#607F4B",  "X" = "blue")) +
          ggplot2::scale_color_manual(values=c('NonSig'='grey85','Sig'='red'))+
          ggplot2::scale_alpha_manual(values=c('NonSig'=0.3,'Sig'=1)) +
          ggplot2::theme(panel.background = ggplot2::element_blank(),
                         strip.background = ggplot2::element_blank(),
                         axis.title.y = ggplot2::element_blank(),
                         axis.title.x = ggplot2::element_blank(),
                         axis.text.x = ggplot2::element_blank(),
                         axis.text.y = ggplot2::element_blank(),
                         axis.ticks.x=ggplot2::element_blank(),
                         axis.ticks.y=ggplot2::element_blank(),
                         legend.position="none")
  )


}
P_Thresh <- 1.02459e-10
draw.pairwise(rbind(G2G_Block  %>% dplyr::filter(P >= P_Thresh),G2G_Block %>% dplyr::filter(P < P_Thresh)),host_chr_length,target_chr_length,'Human','M.tb',1.02459e-10)

#Pathogen Plot
Sig_Hits <- lapply(G2G_Res,function(x) dplyr::filter(x,P <= P_Thresh))
Sig_Hits <- Sig_Hits[sapply(Sig_Hits,function(x) nrow(x) > 0)]
Sig_Hits_Pos <- as.numeric(sapply(names(Sig_Hits),function(x) strsplit(x=x,split = ':')[[1]][2]))

results_dir <- '../../results/Burden_False_SIFT_False_Del_False_HomoOnly_True_HetThresh_10/PLINK/PC_3_pPC_0/Stratified_False/LINEAGE_ALL/'
Mtb_Variants <- dir(results_dir)
Mtb_Variants <- Mtb_Variants[grepl(Mtb_Variants,pattern = 'hybrid.gz')]
Mtb_Variants_Pos <- as.numeric(sapply(Mtb_Variants,function(x) strsplit(x=x,split = ':')[[1]][2]))

library(gggenes)
H37Rv_GFF <- data.table::fread('../../data/Mtb/H37Rv.gff3',skip = 2,sep = '\t') %>% dplyr::select(Type=V3,Start = V4,End=V5,orientation=V7,Annot=V9) %>% dplyr::filter(Type=='gene') %>%
  dplyr::select(-Type) %>% dplyr::mutate(forward = ifelse(orientation == '+',T,F))
H37Rv_GFF$Gene_Name <- sapply(H37Rv_GFF$Annot,function(x) strsplit(strsplit(x=x,split = 'Name=')[[1]][2],split = ';')[[1]][1])
H37Rv_GFF$Locus_Tag <- sapply(H37Rv_GFF$Annot,function(x) strsplit(strsplit(x=x,split = 'locus_tag=')[[1]][2],split = ';')[[1]][1])
H37Rv_GFF <- H37Rv_GFF %>% dplyr::select(-Annot)

Pos_Offset <- 5000
pathogen_plots <- list()
for(i in 1:length(Sig_Hits)){
  Neighbours <- Mtb_Variants_Pos[(Mtb_Variants_Pos > Sig_Hits_Pos[i] - Pos_Offset) & (Mtb_Variants_Pos < Sig_Hits_Pos[i] + Pos_Offset)]
  Neighbours_df <- lapply(Mtb_Variants[match(Neighbours,Mtb_Variants_Pos)],function(x) data.table::fread(cmd = glue::glue('zcat {results_dir}{x}')))
  Neighbours_df <- lapply(Neighbours_df,function(x) dplyr::filter(x,ID == Sig_Hits[[i]]$ID))
  Neighbours_df <- data.frame(POS = Neighbours,P=sapply(Neighbours_df,function(x) x$P),ID = sapply(Mtb_Variants[match(Neighbours,Mtb_Variants_Pos)],function(x) paste0(strsplit(paste0(strsplit(x,split = ':')[[1]][c(1,3)],collapse = '_'),split='\\.')[[1]][1:2],collapse = '.')))
  Neighbours_Genes <- H37Rv_GFF %>% dplyr::filter(Start > Sig_Hits_Pos[i] - Pos_Offset,End < Sig_Hits_Pos[i] + Pos_Offset)
  pathogen_plots[[i]] <- ggplot() +
    geom_gene_arrow(aes(xmin = Start, xmax = End,y=-1.5, fill = Gene_Name,label = Gene_Name,forward = forward),arrowhead_height = unit(10, "mm"), arrowhead_width = unit(3, "mm"),arrow_body_height = unit(10,'mm'),data = Neighbours_Genes) +
    geom_gene_label(aes(xmin = Start, xmax = End,y=-1.5, fill = Gene_Name,label = Gene_Name,forward = forward),align = "left",data = Neighbours_Genes) + geom_point(aes(x=POS,y=-log10(P)),data = Neighbours_df) + ylim(-2,12) +
    labs(y=TeX("$-log_{10}(P)$"),x= 'M.tb Genome Position') + ggtitle(Sig_Hits[[i]]$ID) +
    geom_text_repel(aes(x=POS,y=-log10(P),label = ID),data = dplyr::filter(Neighbours_df,P == min(Neighbours_df$P)))

}

#Host Plot
hg19_GFF <- data.table::fread(cmd = "cat ../../data/Genotyping/gencode.v19.annotation.gff3 | grep -v '#'",sep = '\t') %>% dplyr::select(Chr=V1,Type=V3,Start = V4,End=V5,orientation=V7,Annot=V9) %>% dplyr::filter(Type=='gene') %>%
  dplyr::select(-Type) %>% dplyr::mutate(forward = ifelse(orientation == '+',T,F))
hg19_GFF$Gene_Name <- sapply(hg19_GFF$Annot,function(x) strsplit(strsplit(x=x,split = 'gene_name=')[[1]][2],split = ';')[[1]][1])
hg19_GFF$Gene_Type <- sapply(hg19_GFF$Annot,function(x) strsplit(strsplit(x=x,split = 'gene_type=')[[1]][2],split = ';')[[1]][1])
hg19_GFF <- hg19_GFF %>% dplyr::filter(Gene_Type == 'protein_coding') %>% dplyr::select(-Annot,-Gene_Type)

Pos_Offset <- 200000
host_plots <- list()
for(i in 1:length(Sig_Hits)){
  GWAS_Data <- data.table::fread(cmd = glue::glue('zcat {results_dir}{names(Sig_Hits)[i]}'))
  GWAS_Data <- GWAS_Data %>% dplyr::filter(`#CHROM` == Sig_Hits[[i]]$`#CHROM`) %>%
    dplyr::filter(POS > Sig_Hits[[i]]$POS - Pos_Offset,POS < Sig_Hits[[i]]$POS + Pos_Offset) %>% dplyr::select(POS,ID,P)
  #Get LD Matrix
  system(glue::glue('~/Software/plink --bfile ../../scratch/Burden_False_SIFT_False_Del_False_HomoOnly_True_HetThresh_10/LINEAGE_ALL/TB_DAR_G2G --r2 inter-chr --ld-snp {GWAS_Data$ID[which.min(GWAS_Data$P)]} --ld-window-r2 0 --out ../../scratch/tmp'))
  ld_mat <- data.table::fread('../../scratch/tmp.ld') %>% dplyr::filter(SNP_B %in% GWAS_Data$ID) %>%
    dplyr::select(ID=SNP_B,r2=R2)
  GWAS_Data <- dplyr::left_join(GWAS_Data,ld_mat,by=c('ID'='ID'))
  system('rm ../../scratch/tmp.ld')
  Neighbours_Genes <- hg19_GFF %>% dplyr::filter(Chr == paste0('chr',Sig_Hits[[i]]$`#CHROM`),Start > Sig_Hits[[i]]$POS - Pos_Offset,End < Sig_Hits[[i]]$POS + Pos_Offset)
  host_plots[[i]] <- ggplot() +
    geom_gene_arrow(aes(xmin = Start, xmax = End,y=ifelse(forward,-1.5,-2.5), fill = Gene_Name,label = Gene_Name,forward = forward),arrowhead_height = unit(10, "mm"), arrowhead_width = unit(3, "mm"),arrow_body_height = unit(10,'mm'),data = Neighbours_Genes) +
    geom_gene_label(aes(xmin = Start, xmax = End,y=ifelse(forward,-1.5,-2.5), fill = Gene_Name,label = Gene_Name,forward = forward),align = "left",data = Neighbours_Genes) + geom_point(aes(x=POS,y=-log10(P),color=r2),data = GWAS_Data) + ylim(-3,12) +
    labs(y=TeX("$-log_{10}(P)$"),x= paste0('Human Chr',Sig_Hits[[i]]$`#CHROM`,' Position')) + ggtitle(paste0(strsplit(paste0(strsplit(names(Sig_Hits)[i],split = ':')[[1]][c(1,3)],collapse = '_'),split='\\.')[[1]][1:2],collapse = '.')) +
    geom_text_repel(aes(x=POS,y=-log10(P),label = ID),data = dplyr::filter(GWAS_Data,P == min(GWAS_Data$P)))

}
```

##Phylo Plot
```{r phylo_plot}
library(ggtree)
library(RColorBrewer)
library(ggplot2)
library(ape)

VisualizeG2G <- function(G2G_Obj,host_snp,AA_variant,lineage,phylotree,out_path,sublineage_only = T,tips_to_excl = c()){
  host_genotype <- snpStats::read.plink(bed = G2G_Obj$host_path[lineage],select.snps = host_snp)
  host_dosage <- as(host_genotype$genotypes, Class = 'numeric')
  host_dosage_ordered <- host_dosage[match(G2G_Obj$both_IDs_to_keep[[lineage]]$FAM_ID,rownames(host_dosage)),]
  host_dosage_ordered[host_dosage_ordered==2] <- paste0(host_genotype$map$allele.2,host_genotype$map$allele.2)
  host_dosage_ordered[host_dosage_ordered==1] <- paste0(host_genotype$map$allele.1,host_genotype$map$allele.2)
  host_dosage_ordered[host_dosage_ordered==0] <- paste0(host_genotype$map$allele.1,host_genotype$map$allele.1)
  
  pathogen_dosage <- G2G_Obj$aa_matrix_filt[[lineage]][,AA_variant,drop = F]
  pathogen_dosage[pathogen_dosage == 2] <- 1
  pathogen_dosage_ordered <- pathogen_dosage[match(G2G_Obj$both_IDs_to_keep[[lineage]]$G_NUMBER,rownames(pathogen_dosage)),,drop = F]
  
  metadata <- data.frame(ID = rownames(pathogen_dosage_ordered),Host_SNP = host_dosage_ordered)
  rownames(metadata) <- metadata$ID
  pathogen_data <- data.table::fread('../data/pheno/metadata_Sinergia_final_dataset_human_bac_genome_available_QCed.txt')
  sublineage <- dplyr::filter(pathogen_data,G_NUMBER %in% rownames(pathogen_dosage_ordered)[pathogen_dosage_ordered!=0]) %>% dplyr::select(Sublineage)
  uniq_sublineage <- unique(sublineage$Sublineage)
  
  tree <- ape::read.tree(phylotree)
  if(sublineage_only){
    strains_to_keep <- dplyr::filter(pathogen_data,Sublineage %in% uniq_sublineage)
    tree_filt <- ape::keep.tip(tree,strains_to_keep$G_NUMBER)
  }else{
    tree_filt <- ape::keep.tip(tree,rownames(pathogen_dosage_ordered))
  }

  gg <- ggtree::ggtree(tree_filt,layout='circular',branch.length = 'none')
  gg <- gg %<+% metadata + geom_tippoint(aes(color = factor(Host_SNP)),size=1.5,show.legend = T) + scale_color_manual(values=c('grey','green','red')) + labs(color = host_snp)
  
  nodes <- rep(NA,length(AA_variant))
  colours <- c()
  cl_gradient <- c(brewer.pal(n = 8, name = 'Dark2'),brewer.pal(n = 11, name = 'RdYlBu'))
  names <- c()
  cl_cnt <- 1
  any_clades <- F
  for(i in 1:length(AA_variant)){
    cur_dosage <- as.vector(pathogen_dosage_ordered[,i])
    cur_minor_allele <- as.numeric(names(table(cur_dosage))[which.min(table(cur_dosage))])
    nodes[i] <- getMRCA(tree_filt, as.character(metadata$ID[which(cur_dosage==cur_minor_allele)]))
    #If AA variant is not monophyletic
    if(length(unlist(phangorn::Descendants(tree_filt,nodes[i],type = 'tips'))) != length(which(cur_dosage==cur_minor_allele))){
      #Highlight larget sublineage
      cur_node <- getMRCA(tree_filt, setdiff(as.character(metadata$ID[which(cur_dosage==cur_minor_allele)]),c(pathogen_data$G_NUMBER[!pathogen_data$Sublineage %in% names(sort(table(sublineage$Sublineage),decreasing = T))[1]],tips_to_excl)))
      new_gg <- ggtree::ggtree(tree_filt,layout='circular',branch.length = 'none')
      new_gg <- new_gg %<+% data.frame(metadata,Patho_Variant = factor(as.vector(pathogen_dosage_ordered[,i]))) + geom_tippoint(aes(shape = factor(Patho_Variant),color = factor(Host_SNP)),size=2,show.legend = T) + scale_color_manual(values=c('grey','green','red')) + labs(color = host_snp,shape = AA_variant[i])
      new_gg <- new_gg + geom_hilight(node = cur_node,fill = cl_gradient[cl_cnt],extend = -0.5*cl_cnt,alpha = 0.2)
      system(glue::glue("mkdir -p {out_path}/LINEAGE_{lineage}/"))
      ggsave(filename = glue::glue("{out_path}/LINEAGE_{lineage}/{host_snp}_{AA_variant[i]}.pdf"),plot = new_gg,height = 10,width = 15)
      next
    }
    any_clades <- T
    if(i > 1){
      if(nodes[i] == nodes[i-1]){
        names[cl_cnt] <- paste0(names[cl_cnt],'\n',AA_variant[i])
      }else{
        cl_cnt <- cl_cnt + 1
        names <- c(names,AA_variant[i])
        gg <- gg + geom_hilight(node = nodes[i],fill = cl_gradient[cl_cnt],extend = -0.5*cl_cnt,alpha = 0.2)
        colours <- c(colours,cl_gradient[cl_cnt])
      }
    }else{
      colours <- c(colours,cl_gradient[cl_cnt])
      names <- c(names,AA_variant[i])
      gg <- gg + geom_hilight(node = nodes[i],fill = cl_gradient[cl_cnt],extend = -0.5*cl_cnt,alpha = 0.2)
    }
  }
  if(any_clades){
    # p1 = set_hilight_legend(gg, colours, names) + theme(legend.position="right")
    p1 = gg
    system(glue::glue("mkdir -p {out_path}/LINEAGE_{lineage}/"))
    ggsave(filename = glue::glue("{out_path}/LINEAGE_{lineage}/{host_snp}.pdf"),plot = p1,height = 10,width = 15)
  }
}
VisualizeG2G(G2G_Obj = readRDS('../scratch/Burden_False_SIFT_False_Del_False_HomoOnly_True_HetThresh_10//G2G_Obj.rds'),
             host_snp = 'rs12151990',
             AA_variant = 'Rv2348c_Rv2348c:2626678:p.Ile101Met',
             lineage = 'L4',
             phylotree = '../data/Mtb/RAxML_bestTree.Sinergia_final_dataset_human_bac_genome_available_rerooted.nwk',
             out_path = '../results/Burden_False_SIFT_False_Del_False_HomoOnly_True_HetThresh_10//PLINK/PC_3_pPC_0/Stratified_False/',
             tips_to_excl = 'G36169')

VisualizeG2G(G2G_Obj = readRDS('../scratch/Burden_False_SIFT_False_Del_False_HomoOnly_True_HetThresh_10/G2G_Obj.rds'),
             host_snp = 'rs75769176',
             AA_variant = 'fixA_Rv3029c:3388671:p.Thr67Met',
             lineage = 'L3',
             phylotree = '../data/Mtb/RAxML_bestTree.Sinergia_final_dataset_human_bac_genome_available_rerooted.nwk',
             out_path = '../results/Burden_False_SIFT_False_Del_False_HomoOnly_True_HetThresh_10/PLINK/PC_3_pPC_0/Stratified_False/')
```
##G2G Pheno 
```{r pheno}
CheckPheno <- function(G2G_Obj,host_snp,AA_variant,lineage,pheno){
  host_genotype <- snpStats::read.plink(bed = paste0('../',G2G_Obj$host_path[lineage]),select.snps = host_snp)
  host_dosage <- as(host_genotype$genotypes, Class = 'numeric')
  host_dosage_ordered <- host_dosage[match(G2G_Obj$both_IDs_to_keep[[lineage]]$FAM_ID,rownames(host_dosage)),]
  host_dosage_ordered[host_dosage_ordered==2] <- paste0(host_genotype$map$allele.2,host_genotype$map$allele.2)
  host_dosage_ordered[host_dosage_ordered==1] <- paste0(host_genotype$map$allele.1,host_genotype$map$allele.2)
  host_dosage_ordered[host_dosage_ordered==0] <- paste0(host_genotype$map$allele.1,host_genotype$map$allele.1)
  
  pathogen_dosage <- G2G_Obj$aa_matrix_filt[[lineage]][,AA_variant,drop = F]
  pathogen_dosage[pathogen_dosage == 2] <- 1
  pathogen_dosage_ordered <- pathogen_dosage[match(G2G_Obj$both_IDs_to_keep[[lineage]]$G_NUMBER,rownames(pathogen_dosage)),,drop = F]
  
  metadata <- data.frame(G_NUMBER = rownames(pathogen_dosage_ordered),Host_SNP = host_dosage_ordered,Pathogen_SNP = as.vector(pathogen_dosage_ordered))
  pathogen_data <- data.table::fread('../../data/pheno/metadata_Sinergia_final_dataset_human_bac_genome_available_QCed.txt')
  pathogen_data_jned <- dplyr::left_join(metadata,pathogen_data %>% dplyr::select(PATIENT_ID,G_NUMBER,all_of(pheno)))
  AA_variant_Parsed <- paste0(strsplit(strsplit(AA_variant,split = ':')[[1]][1],split = '_')[[1]][1],'_',strsplit(AA_variant,split = ':')[[1]][3])
  pathogen_data_jned$Group <- paste0(host_snp,':',pathogen_data_jned$Host_SNP,'\n',AA_variant_Parsed,':',ifelse(pathogen_data_jned$Pathogen_SNP==1,'Present','Absent'))
  
  my_comparisons <- expand.grid()

  pathogen_data_filt <- pathogen_data_jned %>% dplyr::filter(Host_SNP != paste0(host_genotype$map$allele.1,host_genotype$map$allele.1) & !is.na(Host_SNP) & !is.na(Pathogen_SNP))
  ggplot2::ggplot(data = pathogen_data_filt,aes(x=Group,y=TB_score)) + geom_boxplot() + stat_n_text() +   stat_compare_means(method = "anova") 

  
  
}
CheckPheno(G2G_Obj = readRDS('../../scratch/Burden_False_SIFT_False_Del_False_HomoOnly_True_HetThresh_10/G2G_Obj.rds'),
             host_snp = 'rs12151990',
             AA_variant = 'Rv2348c_Rv2348c:2626678:p.Ile101Met',
             lineage = 'L4',
             pheno = 'TB_score')

CheckPheno(G2G_Obj = readRDS('../../scratch/Burden_False_SIFT_False_Del_False_HomoOnly_True_HetThresh_10/G2G_Obj.rds'),
             host_snp = 'rs75769176',
             AA_variant = 'fixA_Rv3029c:3388671:p.Thr67Met',
             lineage = 'L3',
             pheno = 'TB_score')


```


##QQ Plot
```{r QQ Plot}
#QQ Plot
Sig_Hits_Sum_Stats <- lapply(names(Sig_Hits),function(x) data.table::fread(glue::glue("zcat {results_dir}{names(Sig_Hits)[i]}")))
qqman::qq(Sig_Hits_Sum_Stats[[1]]$P)
text(paste0('Lambda=',round(GWAS.utils::genomic_inflation(P=Sig_Hits_Sum_Stats[[1]]$P),2)),x=2,y = 8)
qqman::qq(Sig_Hits_Sum_Stats[[2]]$P)
text(paste0('Lambda=',round(GWAS.utils::genomic_inflation(P=Sig_Hits_Sum_Stats[[2]]$P),2)),x=2,y = 8)

result_dir <- '../../results/Burden_False_SIFT_False_Del_False_HomoOnly_True_HetThresh_10/PLINK/PC_3_pPC_0/Stratified_False/LINEAGE_ALL/'
result_files <- dir(result_dir)
all_lambdas <- result_files[sapply(result_files,function(x) grepl(x=x,pattern = 'lambda'))]
lambda_results <- sapply(all_lambdas,function(x) data.table::fread(glue::glue("{result_dir}{x}"))$V2)
hist(lambda_results,main = 'Lambda of All G2G Associations',xlab = '')
```

##Covariates
```{G2G_Cov}
#Summary Statistics (G2G Hits)
bac_load <- data.table::fread('../data/pheno/metadata_Sinergia_final_dataset_human_bac_genome_available_corrected.txt') %>% dplyr::select(PATIENT_ID,Bacterial_load)
G2G_Obj <- readRDS('../scratch/Burden_False_SIFT_False_Del_False_HomoOnly_True_HetThresh_10/G2G_Obj.rds')
AA_Matrix <- as.data.frame(G2G_Obj$aa_matrix_full[,c("Rv2348c_Rv2348c:2626678:p.Ile101Met",'fixA_Rv3029c:3388671:p.Thr67Met')])
AA_Matrix[AA_Matrix==2] <- 1

G2G_Pheno <- cbind(cbind(cbind(cbind(G2G_Obj$covars$ALL %>% dplyr::select(patient_sex,HIV_status,TB_RF_smoking,age,BMI),AA_Matrix),
                   G2G_Obj$host_PCs$ALL %>% dplyr::select(PC1,PC2,PC3)),G2G_Obj$tb_score$ALL %>% dplyr::select(TB_score)),G2G_Obj$xray_score$ALL %>% dplyr::select(Xray_score))
G2G_Pheno$HIV_status[G2G_Pheno$HIV_status=='NA'] <- NA
G2G_Pheno$age <- scale(as.numeric(G2G_Pheno$age))
G2G_Pheno$BMI <- scale(as.numeric(G2G_Pheno$BMI))
G2G_Pheno$TB_RF_smoking <- ifelse(G2G_Pheno$TB_RF_smoking=='yes',1,0)

fixA_mdl_sex <- summary(glm(formula = '`fixA_Rv3029c:3388671:p.Thr67Met`~patient_sex',data = G2G_Pheno))$coefficients['patient_sexmale',c('Estimate','Std. Error','Pr(>|t|)')]
fixA_mdl_age <- summary(glm(formula = '`fixA_Rv3029c:3388671:p.Thr67Met`~ age',data = G2G_Pheno))$coefficients['age',c('Estimate','Std. Error','Pr(>|t|)')]
fixA_mdl_PC <- summary(glm(formula = '`fixA_Rv3029c:3388671:p.Thr67Met`~PC1+PC2+PC3',data = G2G_Pheno))$coefficients[c('PC1','PC2','PC3'),c('Estimate','Std. Error','Pr(>|t|)')]
fixA_mdl_HIV <- summary(glm(formula = '`fixA_Rv3029c:3388671:p.Thr67Met`~HIV_status',data = G2G_Pheno))$coefficients['HIV_statusnegative',c('Estimate','Std. Error','Pr(>|t|)')]
fixA_mdl_Smoking <- summary(glm(formula = '`fixA_Rv3029c:3388671:p.Thr67Met`~TB_RF_smoking',data = G2G_Pheno))$coefficients['TB_RF_smoking',c('Estimate','Std. Error','Pr(>|t|)')]
fixA_mdl_BMI <- summary(glm(formula = '`fixA_Rv3029c:3388671:p.Thr67Met`~BMI',data = G2G_Pheno))$coefficients['BMI',c('Estimate','Std. Error','Pr(>|t|)')]

Rv2348c_mdl_sex <- summary(glm(formula = '`Rv2348c_Rv2348c:2626678:p.Ile101Met`~patient_sex',data = G2G_Pheno))$coefficients['patient_sexmale',c('Estimate','Std. Error','Pr(>|t|)')]
Rv2348c_mdl_age <- summary(glm(formula = '`Rv2348c_Rv2348c:2626678:p.Ile101Met`~ age',data = G2G_Pheno))$coefficients['age',c('Estimate','Std. Error','Pr(>|t|)')]
Rv2348c_mdl_PC <- summary(glm(formula = '`Rv2348c_Rv2348c:2626678:p.Ile101Met`~PC1+PC2+PC3',data = G2G_Pheno))$coefficients[c('PC1','PC2','PC3'),c('Estimate','Std. Error','Pr(>|t|)')]
Rv2348c_mdl_HIV <- summary(glm(formula = '`Rv2348c_Rv2348c:2626678:p.Ile101Met`~HIV_status',data = G2G_Pheno))$coefficients['HIV_statusnegative',c('Estimate','Std. Error','Pr(>|t|)')]
Rv2348c_mdl_Smoking <- summary(glm(formula = '`Rv2348c_Rv2348c:2626678:p.Ile101Met`~TB_RF_smoking',data = G2G_Pheno))$coefficients['TB_RF_smoking',c('Estimate','Std. Error','Pr(>|t|)')]
Rv2348c_mdl_BMI <- summary(glm(formula = '`Rv2348c_Rv2348c:2626678:p.Ile101Met`~BMI',data = G2G_Pheno))$coefficients['BMI',c('Estimate','Std. Error','Pr(>|t|)')]

```

```{r Patient_Characteristics}
#Patient Characteristics (Lineage)
library(table1)
pheno <- data.table::fread('../../data/pheno/metadata_Sinergia_final_dataset_human_bac_genome_available_QCed.txt')
pheno$HIV_status[pheno$HIV_status==""] <- NA


host_PC <- data.table::fread('../../scratch/Host/TB_DAR_GWAS_PCA.eigenvec') %>%
  dplyr::select(IID=V1,PC1=V3,PC2=V4,PC3=V5,PC4=V6)
G2G_Obj <- readRDS('../../scratch/Burden_False_SIFT_False_Del_False_HomoOnly_True_HetThresh_10/G2G_Obj.rds')


host_PC <- host_PC %>% dplyr::inner_join(G2G_Obj$both_IDs_to_keep$ALL,by=c('IID'='FAM_ID')) %>% dplyr::select(PATIENT_ID,PC1,PC2,PC3,PC4)

pheno <- pheno %>% dplyr::left_join(host_PC)

my.render.cont <- function(x) {
  with(stats.apply.rounding(stats.default(x), digits=2), c("",
                                                           "Mean (SD)"=sprintf("%s (&plusmn; %s)", MEAN, SD)))
}
my.render.cat <- function(x) {
  c("", sapply(stats.default(x), function(y) with(y,
                                                  sprintf("%d (%0.0f %%)", FREQ, PCT))))
}
pvalue <- function(x, ...) {
  # Construct vectors of data y, and groups (strata) g
  y <- unlist(x)
  g <- factor(rep(1:length(x), times=sapply(x, length)))
  if (is.numeric(y)) {
    # For numeric variables, perform a standard 2-sample t-test
    p <- round(summary(aov(y ~ g))[[1]][["Pr(>F)"]][1],2)
  } else {
    # For categorical variables, perform a chi-squared test of independence
    p <- round(chisq.test(table(y, g))$p.value,2)
  }
  # Format the p-value, using an HTML entity for the less-than sign.
  # The initial empty string places the output on the line below the variable label.
  c("", sub("<", "&lt;", format.pval(p, digits=3, eps=0.001)))
}

table1(~ factor(patient_sex) + age + BMI + TB_RF_smoking + HIV_status + TB_score + Ct_value + Xray_score | Lineage, 
       data=pheno,render.continuous=my.render.cont, 
       render.categorical=my.render.cat,extra.col=list(`P-value`=pvalue))
```

## PC Plot
```{r PC_Plot}
host_PC_eigenval <- data.table::fread('../../scratch/Host/TB_DAR_GWAS_PCA.eigenval')
ggplot2::ggplot(pheno) + geom_point(aes(x=PC1,y=PC2,color = Lineage)) + 
  labs(x = paste0('Host PC1 (',round(host_PC_eigenval$V1[1] / sum(host_PC_eigenval$V1) * 100,2),'%)'),y=paste0('Host PC2 (',round(host_PC_eigenval$V1[2] / sum(host_PC_eigenval$V1) * 100,2),'%)'))
ggplot2::ggplot(pheno) + geom_point(aes(x=PC3,y=PC4,color = Lineage)) + 
  labs(x = paste0('Host PC3 (',round(host_PC_eigenval$V1[3] / sum(host_PC_eigenval$V1) * 100,2),'%)'),y=paste0('Host PC4 (',round(host_PC_eigenval$V1[4] / sum(host_PC_eigenval$V1) * 100,2),'%)'))
```


## Epitope Circos
```{r circos}
source('../scripts/HLAEnrichment.R')
H37Rv_df <- data.frame(chr = c('H37Rv'),start = 0,end = 4411532)

Epitope_Mutations <- G2G_HLA_Results_Summary %>% dplyr::filter(Is_Epitope)
Epitope_Mutations <- sapply(rownames(Epitope_Mutations),function(x) paste0(strsplit(x=x,split = '\\.')[[1]][1:2],collapse = '.'))
Epitope_Variant_Sets <- lapply(Variant_Set,function(x) x[x %in% Epitope_Mutations])
Epitope_Variant_Sets <- Epitope_Variant_Sets[sapply(Epitope_Variant_Sets,length) > 0]

Epitope_Variant_Sets_Df <- do.call(rbind,lapply(Epitope_Variant_Sets,function(x) {
  pos <- sapply(x,function(y) as.numeric(strsplit(x=y,split = ':')[[1]][2]))
  labels_parsed <- sapply(x,function(y) paste0(strsplit(x=y,split = '_')[[1]][1],' ', strsplit(x=y,split = ':')[[1]][3]))
  return(data.frame(chr = rep('H37Rv',length(x)),start = pos,end = pos,value = labels_parsed))
}))

circos.genomicInitialize(H37Rv_df,axis.labels.cex = 1,tickLabelsStartFromZero = T,major.by = 500000)
plot_cols <- unlist(sapply(1:length(Epitope_Variant_Sets),function(i)rep(brewer.pal(length(Epitope_Variant_Sets),'Paired')[i],length(Epitope_Variant_Sets[[i]]))))
circos.genomicLabels(bed = Epitope_Variant_Sets_Df,col = plot_cols,labels.column = 4, side = "inside")

```
